pkg/pubsub/gcp/publisher.go

```diff
@@ -0,0 +1,154 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcp implements the pubsub.Publisher with Google Cloud Pub/Sub.
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/sigstore/rekor/pkg/events"
+	sigpubsub "github.com/sigstore/rekor/pkg/pubsub"
+
+	"cloud.google.com/go/pubsub"
+	"google.golang.org/api/option"
+)
+
+func init() {
+	sigpubsub.AddProvider(URIIdentifier, func(ctx context.Context, topicResourceID string) (sigpubsub.Publisher, error) {
+		return New(ctx, topicResourceID)
+	})
+}
+
+const URIIdentifier = "gcppubsub://"
+
+var (
+	// Copied from https://github.com/google/go-cloud/blob/master/pubsub/gcppubsub/gcppubsub.go
+	re = regexp.MustCompile(`^gcppubsub://projects/([^/]+)/topics/([^/]+)$`)
+	// Minimal set of permissions needed to check if the server can publish to the configured topic.
+	// https://cloud.google.com/pubsub/docs/access-control#required_permissions
+	requiredIAMPermissions = []string{
+		"pubsub.topics.get",
+		"pubsub.topics.publish",
+	}
+)
+
+type Publisher struct {
+	client *pubsub.Client
+	topic  string
+	wg     *sync.WaitGroup
+}
+
+func New(ctx context.Context, topicResourceID string, opts ...option.ClientOption) (*Publisher, error) {
+	projectID, topic, err := parseRef(topicResourceID)
+	if err != nil {
+		return nil, fmt.Errorf("parse ref: %w", err)
+	}
+	client, err := pubsub.NewClient(ctx, projectID, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("create pubsub client for project %q: %w", projectID, err)
+	}
+
+	// The PubSub emulator does not support IAM methods, and will block the
+	// server start up if they are called. If the environment variable is set,
+	// skip this check.
+	if os.Getenv("PUBSUB_EMULATOR_HOST") == "" {
+		if _, err := client.Topic(topic).IAM().TestPermissions(ctx, requiredIAMPermissions); err != nil {
+			return nil, fmt.Errorf("insufficient permissions for topic %q: %w", topic, err)
+		}
+	}
+
+	return &Publisher{
+		client: client,
+		topic:  topic,
+		wg:     new(sync.WaitGroup),
+	}, nil
+}
+
+func (p *Publisher) Publish(ctx context.Context, event *events.Event, encoding events.EventContentType) error {
+	p.wg.Add(1)
+	defer p.wg.Done()
+
+	var data []byte
+	var err error
+	switch encoding {
+	case events.ContentTypeProtobuf:
+		data, err = event.MarshalProto()
+	case events.ContentTypeJSON:
+		data, err = event.MarshalJSON()
+	default:
+		err = fmt.Errorf("unsupported encoding: %s", encoding)
+	}
+	if err != nil {
+		return fmt.Errorf("marshal event: %w", err)
+	}
+
+	msg := &pubsub.Message{
+		Data:       data,
+		Attributes: gcpAttrs(event, encoding),
+	}
+	res := p.client.Topic(p.topic).Publish(ctx, msg)
+	// TODO: Consider making the timeout configurable.
+	cctx, cancel := context.WithTimeout(ctx, pubsub.DefaultPublishSettings.Timeout)
+	defer cancel()
+	if _, err := res.Get(cctx); err != nil {
```

Comment:
nit: add a comment that explains that `res` is effectively a future, and will block until the prior `Publish` command succeeds?